src/docfx/lib/ParallelUtility.cs

```diff
@@ -51,19 +60,34 @@ public static Task ForEach<T>(IEnumerable<T> source, Func<T, Task> action, Actio
             }
 
             queue.Complete();
-            return queue.Completion;
+
+            try
+            {
+                await queue.Completion;
+            }
+            catch (WrapException we)
+            {
+                ExceptionDispatchInfo.Capture(we.InnerException).Throw();
+            }
 
             async Task Run(T item)
             {
-                await action(item);
+                try
+                {
+                    await action(item);
+                }
+                catch (OperationCanceledException oce)
```

Comment:
>OperationCanceledException [](start = 23, length = 26)

Shall we try a few retry here when we get OperationCanceledException? Especially while we get it with restoring timeout, maybe we do not need to fail the entire run

Reply:
Retry should be handled in the inner action.

Reply:
true

---
In reply to: [226557534](https://github.com/dotnet/docfx/pull/3514#discussion_r226557534) [](ancestors = 226557534)